pkg/syncer/dns/syncer.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dns
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/dynamic/dynamicinformer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/dns/plugin/nsmap"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+)
+
+// StartNsMapSyncer watches for namespace changes scoped by the given informer factory
+// and update the dns nsmap config map
+func StartNsMapSyncer(ctx context.Context, client *dynamic.DynamicClient, informers dynamicinformer.DynamicSharedInformerFactory, ns string) {
+	nsinformer := informers.ForResource(schema.GroupVersionResource{Version: "v1", Resource: "namespaces"})
+	nslister := nsinformer.Lister()
+
+	cmclient := client.Resource(schema.GroupVersionResource{Version: "v1", Resource: "configmaps"}).Namespace(ns)
+
+	nsinformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			updateConfigMap(ctx, cmclient, nslister, ns)
```

Comment:
This must be a proper controller with a work queue and worker goroutines. We can't have event handler functions take lots of time (e.g. make network requests to update a configmap).

Reply:
done. I moved the code to the downstream namespace controller.